---
title: "220908_YoderSC_08_liana"
author: "T.C. Howton"
date: '2022-09-08'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.width=12, fig.height=12) 
```

# Project Overview
This project will use Liana and NichNet to explore cell-cell communication in PKD samples form Dr. Yoder's lab. This project will reproduce Kasi's work with updated NicheNet and adding Liana.

## Script Purpose
LIANA's OmniPath resource was largely compiled from CCC databases with a human focus, as such LIANA works with human gene symbols by default. Nevertheless, one could run LIANA with any other organism, provided a dictionary is available to map human genesymbols to those of species X. This could be achieved using the HomoloGene database via OmnipathR with LIANA's generate_homologs function.


### Load Packages

```{r Load-packages}
suppressPackageStartupMessages({
  library(tidyverse)
  library(liana)
  library(nichenetr)
  library(Seurat)
  library(ggrepel)
  library(cowplot)
  library(OmnipathR)
  library(magrittr)
  library(RColorBrewer)
})
```

###Load Data
```{r Load-Data}
# Loading in the final seurat object from the celltypes script "/data/user/tchowton/220804_YoderSC/Scripts/220822_YoderSC_06_celltypes.Rmd"
alldata <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/220830_YoderSC_06-pc2-final.rds")
alldata <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/220825_YoderSC_06-ift88-final.rds")


dim(alldata) 
# ift88 - 20405 20625
# pc2 - 19893 10688

# I need to add the cell type id of each cell to the meta.data
alldata$celltype <- alldata@active.ident

# Double check meta.data
alldata@meta.data %>% head()

DimPlot(alldata, label = TRUE) + NoAxes() +NoLegend()

DimPlot(alldata, label = FALSE, group.by = "type") + NoAxes()

# check number of cells per cell type per condition
table(alldata@active.ident, alldata@meta.data$type)
dim(alldata)

# change cell type names based on condition origin 
alldata@meta.data$celltype_aggregate = paste(alldata@active.ident, alldata@meta.data$type, sep = "_")

DimPlot(alldata, label = TRUE, group.by = "celltype_aggregate") + NoAxes() + NoLegend()

# change metadata column 
celltype_id <- "celltype_aggregate"
alldata <- SetIdent(alldata, value = alldata[[celltype_id]])
Idents(alldata)

# Prepare object for ccc analysis 
# As LIANA cannot perform differential analyses, I split the Seurat object by condition to be able to apply LIANA to each condition individually. 

alldata_split <- SplitObject(alldata, split.by = "type") 

# check number of cells per cell type per condition
table(alldata_split$Wildtype@active.ident, alldata_split$Wildtype@meta.data$type)

# check number of cells per cell type per condition
table(alldata_split$Mutant@active.ident, alldata_split$Mutant@meta.data$type)

VlnPlot(alldata, idents = c("Resident_Macrophages_Mutant", "Infiltrating_Macrophages_Mutant"), features = "Adgre1")
```

##LIANA

```{r Liana}
# Check the organisms which are currently available in omnipathR.
show_homologene()

# Here, we will convert LIANA's Consensus resource to murine symbols
op_resource <- select_resource("Consensus")[[1]]

# Generate orthologous resource
ortholog_resource <- generate_homologs(op_resource = op_resource,
                                       target_organism = 10090) # mouse

# Run LIANA with the orthologous resource
liana_res_mutant <- liana_wrap(alldata_split$Mutant,
                        resource = 'custom', # resource has to be set to 'custom' to work with external resources
                        external_resource = ortholog_resource, # provide orthologous resource
                        ) %>%
  liana_aggregate()

# Run LIANA with the orthologous resource
liana_res_wildtype <- liana_wrap(alldata_split$Wildtype,
                        resource = 'custom', # resource has to be set to 'custom' to work with external resources
                        external_resource = ortholog_resource, # provide orthologous resource
                        ) %>%
  liana_aggregate()

### Save Liana results pre-filtering ###
#IFT88 - mutant
saveRDS(liana_res_mutant, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221101_ift88_liana-res-prefiltering-mutant.rds")
#IFT88 - Wildtype
saveRDS(liana_res_wildtype, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221101_ift88_liana-res-prefiltering-wildtype.rds")

#PC2 - mutant
saveRDS(liana_res_mutant, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_liana-res-prefiltering-mutant.rds")
#PC2 - Wildtype
saveRDS(liana_res_wildtype, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_liana-res-prefiltering-wildtype.rds")

### Load Liana results pre-filtering ###
#IFT88 - mutant
liana_res_mutant <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221101_ift88_liana-res-prefiltering-mutant.rds")
#IFT88 - wildtype
liana_res_wildtype <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221101_ift88_liana-res-prefiltering-wildtype.rds")

#PC2 - Mutant
liana_res_mutant <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_liana-res-prefiltering-mutant.rds")
#PC2 - wildtype
liana_res_wildtype <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_liana-res-prefiltering-wildtype.rds")

# filter results based on cell types of interest 
liana_res_wildtype <- liana_res_wildtype %>%
  subset(target == c("Proximal_Tubule_S1_Cells_Wildtype")) %>%
  dplyr::rename(ligand = ligand.complex, receptor = receptor.complex) 

liana_res_mutant <- liana_res_mutant %>%
  subset(target == c("Proximal_Tubule_S1_Cells_Mutant")) %>%
  dplyr::rename(ligand = ligand.complex, receptor = receptor.complex) 

# filter for top 50 
n <- 50
top_n_wildtype <- liana_res_wildtype %>%
  arrange(aggregate_rank) %>%
  slice_head(n = n) %>%
  mutate(id = fct_inorder(paste0(source, " - ", ligand, " -> ", receptor)))
top_n_mutant <- liana_res_mutant %>%
  arrange(aggregate_rank) %>%
  slice_head(n = n) %>%
  mutate(id = fct_inorder(paste0(source, " - ", ligand, " -> ", receptor)))

# Visualization of top 50
pdf('/data/user/tchowton/220804_YoderSC/Figures/pc2-nichenet/221206_top-lr-liana-wt.pdf',
    width = 8,
    height = 6)
top_n_wildtype %>%
  ggplot(aes(y = aggregate_rank, x = id)) +
  geom_bar(stat = "identity") +
  xlab("Interaction") + ylab("LIANA's aggregate rank") +
  theme_cowplot() +
  theme(axis.text.x = element_text(size = 8, angle = 60, hjust = 1, vjust = 1))
dev.off()

pdf('/data/user/tchowton/220804_YoderSC/Figures/pc2-nichenet/220929_top-lr-liana-mut.pdf',
    width = 8,
    height = 6)
top_n_mutant %>%
  ggplot(aes(y = aggregate_rank, x = id)) +
  geom_bar(stat = "identity") +
  xlab("Interaction") + ylab("LIANA's aggregate rank") +
  theme_cowplot() +
  theme(axis.text.x = element_text(size = 8, angle = 60, hjust = 1, vjust = 1))
dev.off()
```

### Prioritize ligands from LIANA
```{r}
ligands_wt <- unique(liana_res_wildtype$ligand)
ligands_wt <- ligands_wt[ligands_wt %in% colnames(ligand_target_matrix)]
ligands_wt

ligands_mut <- unique(liana_res_mutant$ligand)
ligands_mut <- ligands_mut[ligands_mut %in% colnames(ligand_target_matrix)]
ligands_mut

ligands <- unique(c(ligands_wt, ligands_mut))

# Save Liana ligands
#PC2
saveRDS(ligands, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_liana-ligands.rds")
#IFT88
saveRDS(ligands, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_ift88_liana-ligands.rds")

#IFT88 - PTC1,2,3 receivers
saveRDS(ligands, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221101_ift88_liana-ligands-ptc123.rds")

#IFT88 - PTC1 receivers
saveRDS(ligands, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221208_ift88_liana-ligands-ptcs1.rds")

# Load Liana Ligands
#PC2 - Res_Macro receiver
ligands <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_liana-ligands.rds")
#IFT88 - Res_Macro receiver
ligands <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_ift88_liana-ligands.rds")

```

##NicheNet
I am using differential nichenet for this analysis.

### Read in NicheNet Priors

```{r Retrieve-Nichenet-Priors}
# Ligand-Target Matrix
ligand_target_matrix = readRDS("/data/user/tchowton/220804_YoderSC/Data/NicheNetPrior/ligand_target_matrix.rds")
#ligand_target_matrix[1:5,1:5] # target genes in rows, ligands in columns


lr_network = readRDS("/data/user/tchowton/220804_YoderSC/Data/NicheNetPrior/lr_network.rds")
lr_network = lr_network %>% mutate(bonafide = ! database %in% c("ppi_prediction","ppi_prediction_go"))
lr_network = lr_network %>% dplyr::rename(ligand = from, receptor = to) %>% distinct(ligand, receptor, bonafide)

#head(lr_network)


# Note: if your data is of mouse origin: convert human gene symbols to their one-to-one orthologs

organism = "mouse" 

if(organism == "mouse"){
  lr_network = lr_network %>% mutate(ligand = convert_human_to_mouse_symbols(ligand), receptor = convert_human_to_mouse_symbols(receptor)) %>% drop_na()

  colnames(ligand_target_matrix) = ligand_target_matrix %>% colnames() %>% convert_human_to_mouse_symbols()
  rownames(ligand_target_matrix) = ligand_target_matrix %>% rownames() %>% convert_human_to_mouse_symbols()
  ligand_target_matrix = ligand_target_matrix %>% .[!is.na(rownames(ligand_target_matrix)), !is.na(colnames(ligand_target_matrix))]
}
```

### Defining Niches

```{r}

### Distal Tubules -> Res Mac ###
niches <- list(
  "pc2_niche" = list(
    "sender" = c("Dital_Convoluted_Tubule_cells_Mutant"),
    "receiver" = c("Resident_Macrophages_Mutant")),
  "wt_niche" = list(
    "sender" = c("Dital_Convoluted_Tubule_cells_Wildtype"),
    "receiver" = c("Resident_Macrophages_Wildtype"))
  ) 

### Res Mac -> Res Mac ###
niches <- list(
  "pc2_niche" = list(
    "sender" = c("Resident_Macrophages_Mutant"),
    "receiver" = c("Resident_Macrophages_Mutant")),
  "wt_niche" = list(
    "sender" = c("Resident_Macrophages_Wildtype"),
    "receiver" = c("Resident_Macrophages_Wildtype"))
  ) 

### Res Mac -> PTC ###
niches <- list(
  "pc2_niche" = list(
    "sender" = c("Resident_Macrophages_Mutant"),
    "receiver" = c("Proximal_Tubule_S1_Cells_Mutant")),
  "wt_niche" = list(
    "sender" = c("Resident_Macrophages_Wildtype"),
    "receiver" = c("Proximal_Tubule_S1_Cells_Wildtype"))
  ) 

### All senders niches ###
niches <- list(
  "pc2_niche" = list(
    "sender" = as.vector(unique(Idents(subset(alldata, subset = type == "Mutant")))[-2]),
    "receiver" = c("Resident_Macrophages_Mutant")),
  "wt_niche" = list(
    "sender" = as.vector(unique(Idents(subset(alldata, subset = type == "Wildtype")))[-6]),
    "receiver" = c("Resident_Macrophages_Wildtype"))
  ) # user adaptation required on own dataset
View(niches)
```




### Calculate differential expression between the niches
In this step, we will determine DE between the different niches for both senders and receivers to define the DE of L-R pairs. The method to calculate the differential expression is here the standard Seurat Wilcoxon test.

```{r}
assay_oi <- "RNA" 

DE_sender <- calculate_niche_de(seurat_obj = alldata %>% subset(features = ligands %>% unique()), niches = niches, type = "sender", assay_oi = assay_oi) # only ligands important for sender cell types

DE_receiver <- calculate_niche_de(seurat_obj = alldata %>% subset(features = lr_network$receptor %>% unique()), niches = niches, type = "receiver", assay_oi = assay_oi) # only receptors now, later on: DE analysis to find targets

DE_sender <- DE_sender %>% mutate(avg_log2FC = ifelse(avg_log2FC == Inf, max(avg_log2FC[is.finite(avg_log2FC)]), ifelse(avg_log2FC == -Inf, min(avg_log2FC[is.finite(avg_log2FC)]), avg_log2FC)))

DE_receiver <- DE_receiver %>% mutate(avg_log2FC = ifelse(avg_log2FC == Inf, max(avg_log2FC[is.finite(avg_log2FC)]), ifelse(avg_log2FC == -Inf, min(avg_log2FC[is.finite(avg_log2FC)]), avg_log2FC)))
```

### Process DE results:

```{r}
expression_pct <- 0.10
DE_sender_processed <- process_niche_de(DE_table = DE_sender, niches = niches, expression_pct = expression_pct, type = "sender")
DE_receiver_processed <- process_niche_de(DE_table = DE_receiver, niches = niches, expression_pct = expression_pct, type = "receiver")

specificity_score_LR_pairs <- "min_lfc"
DE_sender_receiver <- combine_sender_receiver_de(DE_sender_processed, DE_receiver_processed, lr_network, specificity_score = specificity_score_LR_pairs)
```

### Calculate ligand activities and infer active ligand-target links

In this step, we will predict ligand activities of each ligand for each of the receiver cell types across the different niches. This is similar to the ligand activity analysis done in the normal NicheNet pipeline.

To calculate ligand activities, we first need to define a geneset of interest for each niche. 

Note that you can also define these geneset of interest in a different way! (eg pathway-based geneset etc)

Ligand-target links are inferred in the same way as described in the basic NicheNet vignettes.

```{r Predict-ligand-activity}
lfc_cutoff <- 0.25 # because I only have 2 niches 
specificity_score_targets <- "min_lfc"

DE_receiver_targets <- calculate_niche_de_targets(seurat_obj = alldata, niches = niches, lfc_cutoff = lfc_cutoff, expression_pct = expression_pct, assay_oi = assay_oi) 

DE_receiver_processed_targets <- process_receiver_target_de(DE_receiver_targets = DE_receiver_targets, niches = niches, expression_pct = expression_pct, specificity_score = specificity_score_targets)

background <- DE_receiver_processed_targets  %>% pull(target) %>% unique()

geneset_niche1 <- DE_receiver_processed_targets %>% filter(receiver == niches[[1]]$receiver & target_score >= lfc_cutoff & target_significant == 1 & target_present == 1) %>% pull(target) %>% unique()

geneset_niche2 <- DE_receiver_processed_targets %>% filter(receiver == niches[[2]]$receiver & target_score >= lfc_cutoff & target_significant == 1 & target_present == 1) %>% pull(target) %>% unique()
  
geneset_niche1 %>% setdiff(rownames(ligand_target_matrix))

geneset_niche2 %>% setdiff(rownames(ligand_target_matrix))

# You want the geneset length to be between 20-1000
length(geneset_niche1)
## ift88 - 750 | pc2 - 28 | ift88-ptc - 70 | pc2-ptc - 28 | pc2-ptcs1 - 72
length(geneset_niche2)
## ift88 - 181 | pc2 - 56 | ift88-ptc - 30 | pc2-ptc - 56 | pc2-ptcs1 - 30

top_n_target = 250

niche_geneset_list = list(
  "mutant_niche" = list(
    "receiver" = niches[[1]]$receiver,
    "geneset" = geneset_niche1,
    "background" = background),
  "wildtype_niche" = list(
    "receiver" = niches[[2]]$receiver,
    "geneset" = geneset_niche2,
    "background" = background)
  )
  
ligand_activities_targets <- get_ligand_activities_targets(niche_geneset_list = niche_geneset_list, ligand_target_matrix = ligand_target_matrix, top_n_target = top_n_target)

```

### Calculate (scaled) expression of ligands, receptors and targets across cell types of interest (log expression values and expression fractions)

In this step, we will calculate average (scaled) expression, and fraction of expression, of ligands, receptors, and target genes across all cell types of interest. Now this is here demonstrated via the DotPlot function of Seurat, but this can also be done via other ways of course.

```{r}
features_oi <- union(lr_network$ligand, lr_network$receptor)%>% union(ligand_activities_targets$target) %>% setdiff(NA)

dotplot <- suppressWarnings(Seurat::DotPlot(alldata %>% subset(idents = niches %>% unlist() %>% unique()), features = features_oi, assay = assay_oi))

exprs_tbl <- dotplot$data %>% as_tibble()
exprs_tbl <- exprs_tbl %>% rename(celltype = id, gene = features.plot, expression = avg.exp, expression_scaled = avg.exp.scaled, fraction = pct.exp) %>%
    mutate(fraction = fraction/100) %>% as_tibble() %>% select(celltype, gene, expression, expression_scaled, fraction) %>% distinct() %>% arrange(gene) %>% mutate(gene = as.character(gene))
  
exprs_tbl_ligand <- exprs_tbl %>% filter(gene %in% lr_network$ligand) %>% rename(sender = celltype, ligand = gene, ligand_expression = expression, ligand_expression_scaled = expression_scaled, ligand_fraction = fraction) 
exprs_tbl_receptor <- exprs_tbl %>% filter(gene %in% lr_network$receptor) %>% rename(receiver = celltype, receptor = gene, receptor_expression = expression, receptor_expression_scaled = expression_scaled, receptor_fraction = fraction)
exprs_tbl_target <- exprs_tbl %>% filter(gene %in% ligand_activities_targets$target) %>% rename(receiver = celltype, target = gene, target_expression = expression, target_expression_scaled = expression_scaled, target_fraction = fraction)

exprs_tbl_ligand <- exprs_tbl_ligand %>%  mutate(scaled_ligand_expression_scaled = scale_quantile_adapted(ligand_expression_scaled)) %>% mutate(ligand_fraction_adapted = ligand_fraction) %>% mutate_cond(ligand_fraction >= expression_pct, ligand_fraction_adapted = expression_pct)  %>% mutate(scaled_ligand_fraction_adapted = scale_quantile_adapted(ligand_fraction_adapted))

exprs_tbl_receptor <- exprs_tbl_receptor %>% mutate(scaled_receptor_expression_scaled = scale_quantile_adapted(receptor_expression_scaled))  %>% mutate(receptor_fraction_adapted = receptor_fraction) %>% mutate_cond(receptor_fraction >= expression_pct, receptor_fraction_adapted = expression_pct)  %>% mutate(scaled_receptor_fraction_adapted = scale_quantile_adapted(receptor_fraction_adapted))
```

### Expression fraction and receptor

In this step, we will score ligand-receptor interactions based on expression strength of the receptor, in such a way that we give higher scores to the most strongly expressed receptor of a certain ligand, in a certain celltype. This will not effect the rank of individual ligands later on, but will help in prioritizing the most important receptors per ligand (next to other factors regarding the receptor - see later).

```{r}
exprs_sender_receiver <- lr_network %>% 
  inner_join(exprs_tbl_ligand, by = c("ligand")) %>% 
  inner_join(exprs_tbl_receptor, by = c("receptor")) %>% inner_join(DE_sender_receiver %>% distinct(niche, sender, receiver))
  
ligand_scaled_receptor_expression_fraction_df <- exprs_sender_receiver %>% group_by(ligand, receiver) %>% mutate(rank_receptor_expression = dense_rank(receptor_expression), rank_receptor_fraction  = dense_rank(receptor_fraction)) %>% mutate(ligand_scaled_receptor_expression_fraction = 0.5*( (rank_receptor_fraction / max(rank_receptor_fraction)) + ((rank_receptor_expression / max(rank_receptor_expression))) ) )  %>% distinct(ligand, receptor, receiver, ligand_scaled_receptor_expression_fraction, bonafide) %>% distinct() %>% ungroup() 
```

### Adding mock spatial data (differential NicheNet will not run otherwise)
```{r}
include_spatial_info_sender = FALSE 
include_spatial_info_receiver = FALSE
spatial_info <- tibble(celltype_region_oi = "Astrocytes_AD", niche =  "AD_niche", celltype_type = "sender") 
specificity_score_spatial = "lfc"
# this is how this should be defined if you don't have spatial info
# mock spatial info
if(include_spatial_info_sender == FALSE & include_spatial_info_receiver == FALSE){
    spatial_info = tibble(celltype_region_oi = NA, celltype_other_region = NA) %>% mutate(niche =  niches %>% names() %>% head(1), celltype_type = "sender")
} 
# sender spatial info
if(include_spatial_info_sender == TRUE){
  sender_spatial_DE = calculate_spatial_DE(seurat_obj = seurat_obj %>% subset(features = lr_network$ligand %>% unique()), spatial_info = spatial_info %>% filter(celltype_type == "sender"))
  sender_spatial_DE_processed = process_spatial_de(DE_table = sender_spatial_DE, type = "sender", lr_network = lr_network, expression_pct = expression_pct, specificity_score = specificity_score_spatial)
  # add a neutral spatial score for sender celltypes in which the spatial is not known / not of importance
  sender_spatial_DE_others = get_non_spatial_de(niches = niches, spatial_info = spatial_info, type = "sender", lr_network = lr_network)
  sender_spatial_DE_processed = sender_spatial_DE_processed %>% bind_rows(sender_spatial_DE_others)
  sender_spatial_DE_processed = sender_spatial_DE_processed %>% mutate(scaled_ligand_score_spatial = scale_quantile_adapted(ligand_score_spatial))
} else {
  # # add a neutral spatial score for all sender celltypes (for none of them, spatial is relevant in this case)
  sender_spatial_DE_processed = get_non_spatial_de(niches = niches, spatial_info = spatial_info, type = "sender", lr_network = lr_network)
  sender_spatial_DE_processed = sender_spatial_DE_processed %>% mutate(scaled_ligand_score_spatial = scale_quantile_adapted(ligand_score_spatial))  
}
# receiver spatial info 
if(include_spatial_info_receiver == TRUE){
  receiver_spatial_DE = calculate_spatial_DE(seurat_obj = seurat_obj %>% subset(features = lr_network$receptor %>% unique()), spatial_info = spatial_info %>% filter(celltype_type == "receiver"))
  receiver_spatial_DE_processed = process_spatial_de(DE_table = receiver_spatial_DE, type = "receiver", lr_network = lr_network, expression_pct = expression_pct, specificity_score = specificity_score_spatial)
  # add a neutral spatial score for receiver celltypes in which the spatial is not known / not of importance
  receiver_spatial_DE_others = get_non_spatial_de(niches = niches, spatial_info = spatial_info, type = "receiver", lr_network = lr_network)
  receiver_spatial_DE_processed = receiver_spatial_DE_processed %>% bind_rows(receiver_spatial_DE_others)
  receiver_spatial_DE_processed = receiver_spatial_DE_processed %>% mutate(scaled_receptor_score_spatial = scale_quantile_adapted(receptor_score_spatial))
} else {
    # # add a neutral spatial score for all receiver celltypes (for none of them, spatial is relevant in this case)
  receiver_spatial_DE_processed = get_non_spatial_de(niches = niches, spatial_info = spatial_info, type = "receiver", lr_network = lr_network)
  receiver_spatial_DE_processed = receiver_spatial_DE_processed %>% mutate(scaled_receptor_score_spatial = scale_quantile_adapted(receptor_score_spatial))
}
```


### Prioritization of ligand-receptor and ligand-target links

In this step, we will combine all the above calculated information to prioritize ligand-receptor-target links. We scale every property of interest between 0 and 1, and the final prioritization score is a weighted sum of the scaled scores of all the properties of interest.

```{r}
prioritizing_weights <- c("scaled_ligand_score" = 5,
                         "scaled_ligand_expression_scaled" = 1,
                         "ligand_fraction" = 1,
                         "scaled_ligand_score_spatial" = 0, #no spatial data available
                         "scaled_receptor_score" = 0.5,
                         "scaled_receptor_expression_scaled" = 0.5,
                          "receptor_fraction" = 1, 
                         "ligand_scaled_receptor_expression_fraction" = 1,
                         "scaled_receptor_score_spatial" = 0,
                         "scaled_activity" = 0,
                         "scaled_activity_normalized" = 1,
                         "bona_fide" = 1)

output <- list(DE_sender_receiver = DE_sender_receiver, ligand_scaled_receptor_expression_fraction_df = ligand_scaled_receptor_expression_fraction_df, sender_spatial_DE_processed = sender_spatial_DE_processed, receiver_spatial_DE_processed = receiver_spatial_DE_processed, ligand_activities_targets = ligand_activities_targets, DE_receiver_processed_targets = DE_receiver_processed_targets, exprs_tbl_ligand = exprs_tbl_ligand, exprs_tbl_receptor = exprs_tbl_receptor, exprs_tbl_target = exprs_tbl_target)

prioritization_tables <- get_prioritization_tables(output, prioritizing_weights)
```

### Save outputs and prioritization tables
```{r}
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/220929_pc2_prioritiztion-tables.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/220929_pc2_LIANA_NicheNet_output.rds")

# PTC PC2
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_prioritiztion-tables-ptc.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_LIANA_NicheNet_output-ptc.rds")

# PTC IFT88
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_ift88_prioritiztion-tables-ptc.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_ift88_LIANA_NicheNet_output-ptc.rds")

# DTC PC2
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_prioritiztion-tables-dtc.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_LIANA_NicheNet_output-dtc.rds")

# ResMac -> ResMac PC2
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_prioritiztion-tables_ResMac-ResMac.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_LIANA_NicheNet_output_ResMac-ResMac.rds")

# ResMac -> ResMac IFT88
saveRDS(prioritization_tables, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221207_ift88_prioritiztion-tables_ResMac-ResMac.rds")
saveRDS(output, "/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221207_ift88_LIANA_NicheNet_output_ResMac-ResMac.rds")
```


### Load outputs and prioritzation tables
```{r}
alldata <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/220825_YoderSC_06-ift88-final.rds")

prioritization_tables <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_prioritiztion-tables-ptc.rds")
output <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221011_pc2_LIANA_NicheNet_output-ptc.rds")


prioritization_tables <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_prioritiztion-tables-dtc.rds")
output <- readRDS("/data/user/tchowton/220804_YoderSC/Data/ProcessedData/221206_pc2_LIANA_NicheNet_output-dtc.rds")
```

## Visualization of the Differential NicheNet output
### Differential expression of ligand and expression

Before visualization, we need to define the most important ligand-receptor pairs per niche. We will do this by first determining for which niche the highest score is found for each ligand/ligand-receptor pair. And then getting the top 50 ligands per niche.

```{r}
top_ligand_niche_df <- prioritization_tables$prioritization_tbl_ligand_receptor %>% select(niche, sender, receiver, ligand, receptor, prioritization_score) %>% group_by(ligand) %>% top_n(1, prioritization_score) %>% ungroup() %>% select(ligand, receptor, niche) %>% rename(top_niche = niche)

top_ligand_receptor_niche_df <- prioritization_tables$prioritization_tbl_ligand_receptor %>% select(niche, sender, receiver, ligand, receptor, prioritization_score) %>% group_by(ligand, receptor) %>% top_n(1, prioritization_score) %>% ungroup() %>% select(ligand, receptor, niche) %>% rename(top_niche = niche)

ligand_prioritized_tbl_oi <- prioritization_tables$prioritization_tbl_ligand_receptor %>% select(niche, sender, receiver, ligand, prioritization_score) %>% group_by(ligand, niche) %>% top_n(1, prioritization_score) %>% ungroup() %>% distinct() %>% inner_join(top_ligand_niche_df) %>% filter(niche == top_niche) %>% group_by(niche) %>% top_n(50, prioritization_score) %>% ungroup() # get the top50 ligands per niche

# Now we will look first at the top ligand-receptor pairs 

receiver_oi <- "Proximal_Tubule_S1_Cells_Mutant" 

filtered_ligands <- ligand_prioritized_tbl_oi %>% filter(receiver == receiver_oi) %>% pull(ligand) %>% unique()

prioritized_tbl_oi <- prioritization_tables$prioritization_tbl_ligand_receptor %>% filter(ligand %in% filtered_ligands) %>% select(niche, sender, receiver, ligand,  receptor, ligand_receptor, prioritization_score) %>% distinct() %>% inner_join(top_ligand_receptor_niche_df) %>% group_by(ligand) %>% filter(receiver == receiver_oi) %>% top_n(2, prioritization_score) %>% ungroup() 

# Visualization: minimum LFC compared to other niches

lfc_plot <- make_ligand_receptor_lfc_plot(receiver_oi, prioritized_tbl_oi, prioritization_tables$prioritization_tbl_ligand_receptor, plot_legend = FALSE, heights = NULL, widths = NULL)

pdf('/data/user/tchowton/220804_YoderSC/Figures/pc2-nichenet/221206_l-r-lfc-by-celltype_dtc-resmac.pdf',
    width = 10,
    height = 12)
lfc_plot
dev.off()


#Now: visualization of ligand activity and ligand-target links

exprs_activity_target_plot <- make_ligand_activity_target_exprs_plot(receiver_oi, prioritized_tbl_oi,  prioritization_tables$prioritization_tbl_ligand_receptor,  prioritization_tables$prioritization_tbl_ligand_target, output$exprs_tbl_ligand,  output$exprs_tbl_target, lfc_cutoff , ligand_target_matrix, plot_legend = FALSE, heights = NULL, widths = NULL)


pdf('/data/user/tchowton/220804_YoderSC/Figures/pc2-nichenet/221209_l-target-plot_PTCS1-ResMac.pdf',
    width = 16,
    height = 12)
exprs_activity_target_plot$combined_plot
dev.off()

##### Circos Plot #####
filtered_ligands <- ligand_prioritized_tbl_oi %>% filter(receiver == receiver_oi) %>% top_n(10, prioritization_score) %>% pull(ligand) %>% unique()

prioritized_tbl_oi <- prioritization_tables$prioritization_tbl_ligand_receptor %>% filter(ligand %in% filtered_ligands) %>% select(niche, sender, receiver, ligand,  receptor, ligand_receptor, prioritization_score) %>% distinct() %>% inner_join(top_ligand_receptor_niche_df) %>% group_by(ligand) %>% filter(receiver == receiver_oi) %>% top_n(2, prioritization_score) %>% ungroup() 

colors_sender <- brewer.pal(n = prioritized_tbl_oi$sender %>% unique() %>% sort() %>% length(), name = 'Spectral') %>% magrittr::set_names(prioritized_tbl_oi$sender %>% unique() %>% sort())
colors_receiver = c("lavender")  %>% magrittr::set_names(prioritized_tbl_oi$receiver %>% unique() %>% sort())

circos_output = make_circos_lr(prioritized_tbl_oi, colors_sender, colors_receiver)
```


```{r}
filtered_ligands = ligand_prioritized_tbl_oi %>% filter(receiver == receiver_oi) %>% top_n(10, prioritization_score) %>% pull(ligand) %>% unique()

prioritized_tbl_oi = prioritization_tables$prioritization_tbl_ligand_receptor %>% filter(ligand %in% filtered_ligands) %>% select(niche, sender, receiver, ligand,  receptor, ligand_receptor, prioritization_score) %>% distinct() %>% inner_join(top_ligand_receptor_niche_df) %>% group_by(ligand) %>% filter(receiver == receiver_oi) %>% top_n(2, prioritization_score) %>% ungroup() 

colors_sender = brewer.pal(n = prioritized_tbl_oi$sender %>% unique() %>% sort() %>% length(), name = 'Spectral') %>% magrittr::set_names(prioritized_tbl_oi$sender %>% unique() %>% sort())
colors_receiver = c("lavender")  %>% magrittr::set_names(prioritized_tbl_oi$receiver %>% unique() %>% sort())

circos_output = make_circos_lr(prioritized_tbl_oi, colors_sender, colors_receiver)
```

### Testing
```{r}
best_upstream_ligands = prioritized_tbl_oi$ligand %>% unique()

targets_oi <- prioritization_tables$prioritization_tbl_ligand_target %>% filter(target_score >= 0.15) %>% filter(ligand %in% best_upstream_ligands & receiver == receiver_oi) %>% pull(target) %>% unique()

#%>% filter(target_score >= 0.25) %>% filter(ligand %in% best_upstream_ligands & receiver == receiver_oi) %>% pull(target) %>% unique()

targets_oi

pdf('/data/user/tchowton/220804_YoderSC/Figures/pc2-nichenet/221017_spp1-trem2-exp-pc2.pdf',
    width = 8,
    height = 6)
VlnPlot(alldata, features = "Spp1", split.by = "type", idents = c("Resident_Macrophages_Wildtype", "Resident_Macrophages_Mutant"), pt.size = 0)
VlnPlot(alldata, features = "Trem2", split.by = "type", idents = c("Resident_Macrophages_Wildtype", "Resident_Macrophages_Mutant"), pt.size = 0)
dev.off()

VlnPlot(alldata, features = "Agt", split.by = "type", idents = c("Proximal_Tubule_S1_Cells_Mutant", "Proximal_Tubule_S2_Cells_Mutant", "Proximal_Tubule_S3_Cells_Mutant", "Proximal_Tubule_S1_Cells_Wildtype", "Proximal_Tubule_S2_Cells_Wildtype", "Proximal_Tubule_S3_Cells_Wildtype"), pt.size = 0)
```





